### PR TITLE
builtin: add format built-in function

### PR DESCRIPTION
**Pull request by ariesdevil**

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -1589,9 +1589,81 @@ type builtinFormatSig struct {
...
 }
 
+type formatFunc func(float64, int64) string
+
 // See https://dev.mysql.com/doc/refman/5.6/en/string-functions.html#function_format
 func (b *builtinFormatSig) eval(row []types.Datum) (d types.Datum, err error) {
-	return d, errFunctionNotExists.GenByArgs("format")
+	args, err := b.evalArgs(row)
+	if err != nil {
+		return types.Datum{}, errors.Trace(err)
```
Comment (shenli) [inline expression/builtin_string.go]: return d, errors.Trace(err)

File: expression/builtin_string.go [hunk 2]
```diff
@@ -1589,9 +1589,81 @@ type builtinFormatSig struct {
...
+		return d, errors.Trace(err)
+	}
+	arg1, err := args[1].ToInt64(sc)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	var arg2 string
+
+	// TODO: Neet to be moved away.
+	supportedFormat := map[string]formatFunc{
```
Comment (shenli) [inline expression/builtin_string.go]: You can define some consts in mysql/const.go.

File: expression/builtin_string.go [hunk 3]
```diff
@@ -1589,9 +1589,81 @@ type builtinFormatSig struct {
...
+	if !exist {
+		arg2 = "en-US"
+	}
+	number, err := arg0.ToFloat64()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+
+	d.SetString(supportedFormat[arg2](number, arg1))
+	return d, errors.Trace(err)
```
Comment (shenli) [inline expression/builtin_string.go]: return d, nil

File: expression/builtin_string.go [hunk 4]
```diff
@@ -1589,9 +1589,81 @@ type builtinFormatSig struct {
...
+}
+
+// TODO: Move those functions away
+func _formatENUS(number float64, precision int64) string {
```
Comment (shenli) [inline expression/builtin_string.go]: The function name is not in golang style. Rename it to `formatENUS`.

File: expression/builtin_string.go [hunk 5]
```diff
@@ -1589,9 +1589,81 @@ type builtinFormatSig struct {
...
+		} else {
+			buffer.WriteString(parts[1])
+			buffer.WriteString(strings.Repeat("0", int(precision)-len(parts[1])))
+		}
+	}
+	return buffer.String()
+}
+
+func _formatZHCN(number float64, precision int64) string {
+	// TODO
```
Comment (shenli) [inline expression/builtin_string.go]: Should we return error here?

File: expression/builtin_string.go [hunk 6]
```diff
@@ -1589,9 +1589,81 @@ type builtinFormatSig struct {
...
+	arg1, err := args[1].ToInt64(sc)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	var arg2 string
+
+	// TODO: Neet to be moved away.
+	supportedFormat := map[string]formatFunc{
+		"en-US": _formatENUS,
+		"zh-CN": _formatZHCN,
```
Comment (tiancaiamao) [inline expression/builtin_string.go]: _ is seldom used in golang for naming

File: expression/builtin_string.go [hunk 7]
```diff
@@ -1589,9 +1589,81 @@ type builtinFormatSig struct {
...
+	}
+	if len(args) == 3 {
+		arg2, err = args[2].ToString()
+	}
+	_, exist := supportedFormat[arg2]
+	if !exist {
+		arg2 = "en-US"
+	}
+	number, err := arg0.ToFloat64()
```
Comment (tiancaiamao) [inline expression/builtin_string.go]: Will this operation loss precision? `SELECT FORMAT(12332.1234567890123456789012345678901, 22);`
PTAL @coocood

File: expression/builtin_string.go [hunk 8]
```diff
@@ -1589,9 +1589,81 @@ type builtinFormatSig struct {
...
+
+// TODO: Move those functions away
+func _formatENUS(number float64, precision int64) string {
+	var buffer bytes.Buffer
+	if number < 0 {
+		buffer.Write([]byte{'-'})
+		number = 0 - number
+	}
+	comma := []byte{','}
+	parts := strings.Split(strconv.FormatFloat(number, 'f', -1, 64), ".")
```
Comment (tiancaiamao) [inline expression/builtin_string.go]: check `len(parts) == 2`  ?
if you strconv.format(1), maybe no decimal part there, I guess...

Comment (tiancaiamao) [inline expression/builtin_string.go]: ok,  I see line 1652...

File: expression/builtin_string_test.go [hunk 9]
```diff
@@ -1054,3 +1054,47 @@ func (s *testEvaluatorSuite) TestOct(c *C) {
...
+		precision interface{}
+		ret       string
+	}{
+		{12332.123456, 4, "12,332.1234"},
+		{12332.123456, 0, "12,332"},
+		{12332.123456, -4, "12,332"},
+		{-12332.123456, 4, "-12,332.1234"},
+		{-12332.123456, 0, "-12,332"},
+		{-12332.123456, -4, "-12,332"},
+		{"-12332.123456", "4", "-12,332.1234"},
```
Comment (tiancaiamao) [inline expression/builtin_string_test.go]: add case that decimal part is very long
add case no decimal part

Comment (tiancaiamao) [inline expression/builtin_string_test.go]: and test the invalid input such as NULL or string and other type, check mysql behavior to confirm

File: expression/builtin_string.go (hunk 7 above)
Comment (coocood) [inline expression/builtin_string.go]: Yes.

**Comment by ariesdevil** [comment]
@tiancaiamao @shenli @coocood please take a look again.

File: expression/builtin_string_test.go [hunk 10]
```diff
@@ -1099,6 +1099,78 @@ func (s *testEvaluatorSuite) TestOct(c *C) {
...
+		ret       interface{}
+	}{
+		{12332.123456, 4, "12,332.1234"},
+		{12332.123456, 0, "12,332"},
+		{12332.123456, -4, "12,332"},
+		{-12332.123456, 4, "-12,332.1234"},
+		{-12332.123456, 0, "-12,332"},
+		{-12332.123456, -4, "-12,332"},
+		{"-12332.123456", "4", "-12,332.1234"},
+		{"12332.1234567890123456789012345678901", 22, "12,332.1234567890123456789012"},
```
Comment (tiancaiamao) [inline expression/builtin_string_test.go]: Hi, @ariesdevil 

I mean you should also handle this case in your branch:
```
mysql> select format("abcd", 32);
ERROR 2013 (HY000): Lost connection to MySQL server during query
```

Comment (ariesdevil) [inline expression/builtin_string_test.go]: @tiancaiamao soon

**Comment by ariesdevil** [comment]
@tiancaiamao Add more tests and please take a look again.

File: expression/builtin_string_test.go [hunk 11]
```diff
@@ -1171,6 +1186,46 @@ func (s *testEvaluatorSuite) TestFormat(c *C) {
...
+		{[]interface{}{"Quadratic", -1, 4, "What"}, "Quadratic"},
+		{[]interface{}{"Quadratic", 3, 100, "What"}, "QuWhat"},
+		{[]interface{}{nil, 3, 100, "What"}, nil},
+		{[]interface{}{"Quadratic", nil, 4, "What"}, nil},
+		{[]interface{}{"Quadratic", 3, nil, "What"}, nil},
+		{[]interface{}{"Quadratic", 3, 4, nil}, nil},
+		{[]interface{}{"Quadratic", 3, -1, "What"}, "QuWhat"},
+		{[]interface{}{"Quadratic", 3, 1, "What"}, "QuWhatdratic"},
+
+		{[]interface{}{"我叫小雨呀", 3, 2, "王雨叶"}, "我叫王雨叶呀"},
```
Comment (tiancaiamao) [inline expression/builtin_string_test.go]: Who is "王雨叶" ?  Foreigners will feel 干(yi)的(lian)漂(meng)亮(bi) when they see this.
TiDB is a global project, it should not use those meaningless string.

Comment (ariesdevil) [inline expression/builtin_string_test.go]: @tiancaiamao It's not my code.  That's `TestInsert`, not `TestFormat`.

Comment (ariesdevil) [inline expression/builtin_string_test.go]: @tiancaiamao The Insert function was implemented by @hiwjd , PR: https://github.com/pingcap/tidb/issues/2863

**Comment by tiancaiamao** [comment]
LGTM
@shenli @coocood

**Pull request merged**